File: server/src/main/java/gg/mineral/server/network/packet/play/clientbound/OpenWindowPacket.java
package gg.mineral.server.network.packet.play.clientbound;

import gg.mineral.api.network.packet.Packet;

import io.netty.buffer.ByteBuf;

public final record OpenWindowPacket(short windowId, short inventoryType, String windowTitle, short numberOfSlots,
        boolean useProvidedWindowTitle, int entityId) implements Packet.OUTGOING {
    @Override
    public void serialize(ByteBuf os) {
        os.writeByte(windowId);
        os.writeByte(inventoryType);
        writeString(os, windowTitle);
        os.writeByte(numberOfSlots);
        os.writeBoolean(useProvidedWindowTitle);
        os.writeInt(entityId);
    }

    @Override
    public byte getId() {
        return 0x2D;
    }
}

File: server/src/main/kotlin/gg/mineral/server/network/packet/play/clientbound/OpenWindowPacket.kt
package gg.mineral.server.network.packet.play.clientbound

import gg.mineral.api.network.packet.Packet
import io.netty.buffer.ByteBuf

@JvmRecord
data class OpenWindowPacket(
    val windowId: UByte, val inventoryType: UByte, val windowTitle: String, val numberOfSlots: UByte,
    val useProvidedWindowTitle: Boolean, val entityId: Int
) : Packet.Outgoing {
    @OptIn(ExperimentalUnsignedTypes::class)
    override fun serialize(os: ByteBuf) {
        os.writeByte(windowId, inventoryType)
        os.writeString(windowTitle)
        os.writeByte(numberOfSlots)
        os.writeBoolean(useProvidedWindowTitle)
        os.writeInt(entityId)
    }

    override val id: Byte
        get() = 0x2D
}
